lib/pointing_party_web/channels/room_channel.ex

```diff
@@ -19,6 +20,103 @@ defmodule PointingPartyWeb.RoomChannel do
   def handle_in("user_estimated", %{"points" => points}, socket) do
     Presence.update(socket, socket.assigns.username, &(Map.put(&1, :points, points)))
 
+    if everyone_voted?(socket) do
+      finalize_voting(socket)
+    end
+
     {:noreply, socket}
   end
+
+  def handle_in("finalized_points", %{"points" => points}, socket) do
+    updated_socket = save_vote_next_card(points, socket)
+    broadcast!(updated_socket, "new_card", %{card: current_card(updated_socket)})
+    {:reply, :ok, updated_socket}
+  end
+
+  def handle_in("start_pointing", _params, socket) do
+    updated_socket = initialize_state(socket)
+    broadcast!(updated_socket, "new_card", %{card: current_card(updated_socket)})
+    {:reply, :ok, updated_socket}
+  end
+
+  defp current_card(socket) do
+    socket.assigns
+    |> Map.get(:current)
+    |> Map.from_struct()
+    |> Map.drop([:__meta__])
+  end
+
+  defp everyone_voted?(socket) do
+    socket
+    |> Presence.list()
+    |> Enum.map(fn {_username, %{metas: [metas]}} -> Map.get(metas, :points) end)
+    |> Enum.all?(&(not is_nil(&1)))
+  end
+
+  defp finalize_voting(socket) do
+    current_users = Presence.list(socket)
+
+    {event, points} =
+      case winning_vote(current_users) do
+        top_two when is_list(top_two) -> {"tie", top_two}
+        winner -> {"winner", winner}
+      end
+
+    broadcast!(socket, event, %{points: points})
+  end
+
+  defp initialize_state(%{assigns: %{cards: _cards}} = socket), do: socket
+  defp initialize_state(socket) do
+    [first | cards] = Card.cards()
+
+    socket
+    |> assign(:points, Card.points_range())
+    |> assign(:unvoted, cards)
+    |> assign(:current, first)
+  end
+
+  defp save_vote_next_card(points, socket) do
+    latest_card =
+      socket.assigns
+      |> Map.get(:current)
+      |> Map.put(:points, points)
+
+    {next, remaining} =
+      socket.assigns
+      |> Map.get(:unvoted)
+      |> List.pop_at(0)
+
+    socket
+    |> assign(:unvoted, remaining)
+    |> assign(:current, next)
+    |> assign(:voted, [latest_card | socket.assigns[:voted]])
+  end
+
+  defp winning_vote(users) do
```

Comment:
There is probably a _much_ easier way to do this 😆

Reply:
Might be a good idea to at least move this business logic out of the channel and into a core module? That way easier to refactor to live view too.

Reply:
I'm good merging this in for now and doing that in a separate PR though.

Reply:
I can do a follow-up PR 👍